### ISSUE TITLE
[C#] Implement new Concept Exercise: memory-pooling

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/csharp#1446]
This issue describes how to implement the `memory-pooling` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student advanced aspects of the Concept of Memory Pooling in C#.

## Learning objectives

- Know what memory pooling is and why it should be used.
- Know how to use the `ArrayPool<T>` and `MemoryPool<T>` classes.

## Out of scope

- Discussion of profiling tools (referring to them is fine).

## Concepts

This Concepts Exercise's Concepts are:

- `memory-pooling`: know what memory pooling is and why it should be used; know how to use the `ArrayPool<T>` and `MemoryPool<T>` classes.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `memory-allocation`: know the basics of how memory works in C#, including reference and value type memory and garbage collector aspects.
- `generic-types`: know how to work with generic types.

### Hints

- [`ArrayPool<T>` type][array-pool]: `ArrayPool<T>` reference.
- [`MemoryPool<T>` type][memory-pool]: `MemoryPool<T>` reference.

### After

- [`ArrayPool<T>` type][array-pool]: `ArrayPool<T>` reference.
- [`MemoryPool<T>` type][memory-pool]: `MemoryPool<T>` reference.
- [ArrayPool in-depth][array-pool-in-depth]: a very in-depth discussion of how to work with ArrayPools.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

To implement this exercise, please [follow these instructions](https://github.com/exercism/v3/blob/master/languages/csharp/reference/implementing-a-concept-exercise.md).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/concepts/memory-pooling.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[array-pool]: https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1?view=netcore-3.1
[memory-pool]: https://docs.microsoft.com/en-us/dotnet/api/system.buffers.memorypool-1?view=netcore-3.1
[array-pool-in-depth]: https://adamsitnik.com/Array-Pool/
